analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/ProgramStateTest.cs

```diff
@@ -290,6 +291,54 @@ public void ToString_WithAll()
 ");
         }
 
+        [TestMethod]
+        public void ResetFieldConstraints_NotPreservesField() =>
+            ResetFieldConstraintTests(PreserveOnFieldResetConstraint.AlwaysReset, shouldBePreserved: false);
+
+        [TestMethod]
+        public void ResetFieldConstraints_PreservesField() =>
+            ResetFieldConstraintTests(PreserveOnFieldResetConstraint.AlwaysPreserve, shouldBePreserved: true);
+
+        [TestMethod]
+        public void ResetFieldConstraints_Default() =>
+            ResetFieldConstraintTests(PreserveOnFieldResetConstraint.Default, shouldBePreserved: false);
+
+        [TestMethod]
+        public void ResetFieldConstraints_ResetFieldsAsDefined()
+        {
+            var instanceField = TestHelper.GetFieldSymbol("object field;");
+            var staticField = TestHelper.GetFieldSymbol("static object field;");
+
+            var preserveInstance = PreserveOnFieldResetConstraint.DistingushConstraint<bool>(preserveOnFieldReset: x => !x.IsStatic);
+            var preserveAll = PreserveOnFieldResetConstraint.DistingushConstraint<byte>(preserveOnFieldReset: _ => true);
+            var preserveNone = PreserveOnFieldResetConstraint.DistingushConstraint<int>(preserveOnFieldReset: _ => false);
+
+            var sut = ProgramState.Empty;
+            sut = sut.SetSymbolValue(instanceField, new SymbolicValue().WithConstraint(preserveInstance).WithConstraint(preserveAll).WithConstraint(preserveNone))
+                     .SetSymbolValue(staticField, new SymbolicValue().WithConstraint(preserveInstance).WithConstraint(preserveAll).WithConstraint(preserveNone));
+            sut = sut.ResetFieldConstraints();
+            var instanceFieldSymbolValue = sut[instanceField];
+            var staticFieldSymbolValue = sut[staticField];
+            instanceFieldSymbolValue.HasConstraint(preserveInstance).Should().BeTrue(because: "preserveInstance should be preserved for instance fields");
+            instanceFieldSymbolValue.HasConstraint(preserveAll).Should().BeTrue(because: "preserveAll should be preserved for instance fields");
+            instanceFieldSymbolValue.HasConstraint(preserveNone).Should().BeFalse(because: "preserveNone should not be preserved for instance fields");
+            staticFieldSymbolValue.HasConstraint(preserveInstance).Should().BeFalse(because: "preserveInstance should not be preserved for static fields");
+            staticFieldSymbolValue.HasConstraint(preserveAll).Should().BeTrue(because: "preserveAll should be preserved for static fields");
+            staticFieldSymbolValue.HasConstraint(preserveNone).Should().BeFalse(because: "preserveNone should not be preserved for static fields");
+        }
+
+        private static void ResetFieldConstraintTests(PreserveOnFieldResetConstraint constraint, bool shouldBePreserved)
```

Comment:
The variable should start with `expected` word as it's used for assertion. I don't have very good idea for the rest then.
```suggestion
        private static void ResetFieldConstraintTests(PreserveOnFieldResetConstraint constraint, bool expectedIsPreserved)
```